File: src/astro.rs
use chrono::{DateTime, NaiveDate, NaiveDateTime, NaiveTime, TimeZone, Utc};

use crate::types::{SolarTerm, TermDef};

pub const TERM_DEFS: [TermDef; 24] = [
    TermDef {
        key: "xiaohan",
        name_ko: "소한",
        name_hanja: "小寒",
        name_en: "Xiaohan",
        angle: 285.0,
    },
    TermDef {
        key: "dahan",
        name_ko: "대한",
        name_hanja: "大寒",
        name_en: "Dahan",
        angle: 300.0,
    },
    TermDef {
        key: "lichun",
        name_ko: "입춘",
        name_hanja: "立春",
        name_en: "Lichun",
        angle: 315.0,
    },
    TermDef {
        key: "yushui",
        name_ko: "우수",
        name_hanja: "雨水",
        name_en: "Yushui",
        angle: 330.0,
    },
    TermDef {
        key: "jingzhe",
        name_ko: "경칩",
        name_hanja: "驚蟄",
        name_en: "Jingzhe",
        angle: 345.0,
    },
    TermDef {
        key: "chunfen",
        name_ko: "춘분",
        name_hanja: "春分",
        name_en: "Chunfen",
        angle: 0.0,
    },
    TermDef {
        key: "qingming",
        name_ko: "청명",
        name_hanja: "清明",
        name_en: "Qingming",
        angle: 15.0,
    },
    TermDef {
        key: "guyu",
        name_ko: "곡우",
        name_hanja: "谷雨",
        name_en: "Guyu",
        angle: 30.0,
    },
    TermDef {
        key: "lixia",
        name_ko: "입하",
        name_hanja: "立夏",
        name_en: "Lixia",
        angle: 45.0,
    },
    TermDef {
        key: "xiaoman",
        name_ko: "소만",
        name_hanja: "小滿",
        name_en: "Xiaoman",
        angle: 60.0,
    },
    TermDef {
        key: "mangzhong",
        name_ko: "망종",
        name_hanja: "芒種",
        name_en: "Mangzhong",
        angle: 75.0,
    },
    TermDef {
        key: "xiazhi",
        name_ko: "하지",
        name_hanja: "夏至",
        name_en: "Xiazhi",
        angle: 90.0,
    },
    TermDef {
        key: "xiaoshu",
        name_ko: "소서",
        name_hanja: "小暑",
        name_en: "Xiaoshu",
        angle: 105.0,
    },
    TermDef {
        key: "dashu",
        name_ko: "대서",
        name_hanja: "大暑",
        name_en: "Dashu",
        angle: 120.0,
    },
    TermDef {
        key: "liqiu",
        name_ko: "입추",
        name_hanja: "立秋",
        name_en: "Liqiu",
        angle: 135.0,
    },
    TermDef {
        key: "chushu",
        name_ko: "처서",
        name_hanja: "處暑",
        name_en: "Chushu",
        angle: 150.0,
    },
    TermDef {
        key: "bailu",
        name_ko: "백로",
        name_hanja: "白露",
        name_en: "Bailu",
        angle: 165.0,
    },
    TermDef {
        key: "qiufen",
        name_ko: "추분",
        name_hanja: "秋分",
        name_en: "Qiufen",
        angle: 180.0,
    },
    TermDef {
        key: "hanlu",
        name_ko: "한로",
        name_hanja: "寒露",
        name_en: "Hanlu",
        angle: 195.0,
    },
    TermDef {
        key: "shuangjiang",
        name_ko: "상강",
        name_hanja: "霜降",
        name_en: "Shuangjiang",
        angle: 210.0,
    },
    TermDef {
        key: "lidong",
        name_ko: "입동",
        name_hanja: "立冬",
        name_en: "Lidong",
        angle: 225.0,
    },
    TermDef {
        key: "xiaoxue",
        name_ko: "소설",
        name_hanja: "小雪",
        name_en: "Xiaoxue",
        angle: 240.0,
    },
    TermDef {
        key: "daxue",
        name_ko: "대설",
        name_hanja: "大雪",
        name_en: "Daxue",
        angle: 255.0,
    },
    TermDef {
        key: "dongzhi",
        name_ko: "동지",
        name_hanja: "冬至",
        name_en: "Dongzhi",
        angle: 270.0,
    },
];

pub fn jd_from_datetime(dt: DateTime<Utc>) -> f64 {
    let seconds = dt.timestamp() as f64 + (dt.timestamp_subsec_nanos() as f64) / 1e9;
    seconds / 86400.0 + 2440587.5
}

pub fn datetime_from_jd(jd: f64) -> DateTime<Utc> {
    let seconds = (jd - 2440587.5) * 86400.0;
    let mut whole = seconds.floor() as i64;
    let mut nanos = ((seconds - whole as f64) * 1e9).round() as i64;
    if nanos >= 1_000_000_000 {
        whole += 1;
        nanos -= 1_000_000_000;
    } else if nanos < 0 {
        whole -= 1;
        nanos += 1_000_000_000;
    }
    DateTime::<Utc>::from_timestamp(whole, nanos as u32).unwrap()
}

pub fn jd_from_utc_date(year: i32, month: u32, day: u32, hour: u32, min: u32, sec: u32) -> f64 {
    let date = NaiveDate::from_ymd_opt(year, month, day).unwrap();
    let time = NaiveTime::from_hms_opt(hour, min, sec).unwrap();
    let naive = NaiveDateTime::new(date, time);
    let dt = Utc.from_utc_datetime(&naive);
    jd_from_datetime(dt)
}

pub fn compute_solar_terms(year: i32) -> Vec<SolarTerm> {
    let start = jd_from_utc_date(year, 1, 1, 0, 0, 0);
    let end = jd_from_utc_date(year + 1, 1, 1, 0, 0, 0);
    let days = (end - start).ceil() as i64;

    let mut targets = Vec::with_capacity(TERM_DEFS.len());
    let mut last = -1.0;
    for def in TERM_DEFS.iter() {
        let mut angle = def.angle;
        while angle <= last {
            angle += 360.0;
        }
        targets.push(angle);
        last = angle;
    }

    let mut results = Vec::with_capacity(TERM_DEFS.len());
    let mut target_idx = 0;
    let mut prev_jd = start;
    let mut prev_unwrapped = sun_apparent_longitude(prev_jd);

    for day in 1..=days {
        let jd = start + day as f64;
        let mut lon = sun_apparent_longitude(jd);
        if lon < prev_unwrapped {
            lon += 360.0;
        }
        while target_idx < targets.len() && targets[target_idx] <= lon {
            let target = targets[target_idx];
            if target < prev_unwrapped {
                target_idx += 1;
                continue;
            }
            let term_jd = refine_term(prev_jd, jd, prev_unwrapped, target);
            results.push(SolarTerm {
                def: &TERM_DEFS[target_idx],
                jd: term_jd,
            });
            target_idx += 1;
        }
        prev_jd = jd;
        prev_unwrapped = lon;
    }

    results
}

fn sun_apparent_longitude(jd: f64) -> f64 {
    let t = (jd - 2451545.0) / 36525.0;
    let l0 = 280.46646 + 36000.76983 * t + 0.0003032 * t * t;
    let m = 357.52911 + 35999.05029 * t - 0.0001537 * t * t;
    let m_rad = deg_to_rad(m);
    let c = (1.914602 - 0.004817 * t - 0.000014 * t * t) * m_rad.sin()
        + (0.019993 - 0.000101 * t) * (2.0 * m_rad).sin()
        + 0.000289 * (3.0 * m_rad).sin();
    let true_long = l0 + c;
    let omega = 125.04 - 1934.136 * t;
    let lambda = true_long - 0.00569 - 0.00478 * deg_to_rad(omega).sin();
    norm_deg(lambda)
}

fn refine_term(jd0: f64, jd1: f64, lon0: f64, target: f64) -> f64 {
    let mut lo = jd0;
    let mut hi = jd1;
    let mut lo_lon = lon0;
    for _ in 0..60 {
        let mid = (lo + hi) / 2.0;
        let mut mid_lon = sun_apparent_longitude(mid);
        if mid_lon < lo_lon {
            mid_lon += 360.0;
        }
        if mid_lon >= target {
            hi = mid;
        } else {
            lo = mid;
            lo_lon = mid_lon;
        }
    }
    (lo + hi) / 2.0
}

fn deg_to_rad(deg: f64) -> f64 {
    deg.to_radians()
}

fn norm_deg(mut deg: f64) -> f64 {
    deg = deg % 360.0;
    if deg < 0.0 {
        deg += 360.0;
    }
    deg
}

File: packages/saju-lib/src/astro.ts
/**
 * @fileoverview 천문 계산 모듈 — 절기(節氣) 시점 산출
 *
 * 태양의 시황경(apparent longitude)을 계산하여 24절기의 정확한 발생 시점을
 * 율리우스일(Julian Date)로 산출한다. 사주 연주·월주 결정에 핵심적인 모듈이다.
 *
 * 태양 위치 계산에는 VSOP87 이론의 간략 공식(low-precision solar position)을
 * 사용한다. Jean Meeus, "Astronomical Algorithms" 참조.
 */

import type { SolarTerm, TermDef } from './types.js';

/**
 * 24절기 정의 배열.
 * 소한(285°)부터 동지(270°)까지 황경 순서로 배열된다.
 * angle은 춘분점(0°) 기준 태양 황경이다.
 */
export const TERM_DEFS: TermDef[] = [
  { key: 'xiaohan', nameKo: '소한', nameHanja: '小寒', nameEn: 'Xiaohan', angle: 285.0 },
  { key: 'dahan', nameKo: '대한', nameHanja: '大寒', nameEn: 'Dahan', angle: 300.0 },
  { key: 'lichun', nameKo: '입춘', nameHanja: '立春', nameEn: 'Lichun', angle: 315.0 },
  { key: 'yushui', nameKo: '우수', nameHanja: '雨水', nameEn: 'Yushui', angle: 330.0 },
  { key: 'jingzhe', nameKo: '경칩', nameHanja: '驚蟄', nameEn: 'Jingzhe', angle: 345.0 },
  { key: 'chunfen', nameKo: '춘분', nameHanja: '春分', nameEn: 'Chunfen', angle: 0.0 },
  { key: 'qingming', nameKo: '청명', nameHanja: '清明', nameEn: 'Qingming', angle: 15.0 },
  { key: 'guyu', nameKo: '곡우', nameHanja: '谷雨', nameEn: 'Guyu', angle: 30.0 },
  { key: 'lixia', nameKo: '입하', nameHanja: '立夏', nameEn: 'Lixia', angle: 45.0 },
  { key: 'xiaoman', nameKo: '소만', nameHanja: '小滿', nameEn: 'Xiaoman', angle: 60.0 },
  { key: 'mangzhong', nameKo: '망종', nameHanja: '芒種', nameEn: 'Mangzhong', angle: 75.0 },
  { key: 'xiazhi', nameKo: '하지', nameHanja: '夏至', nameEn: 'Xiazhi', angle: 90.0 },
  { key: 'xiaoshu', nameKo: '소서', nameHanja: '小暑', nameEn: 'Xiaoshu', angle: 105.0 },
  { key: 'dashu', nameKo: '대서', nameHanja: '大暑', nameEn: 'Dashu', angle: 120.0 },
  { key: 'liqiu', nameKo: '입추', nameHanja: '立秋', nameEn: 'Liqiu', angle: 135.0 },
  { key: 'chushu', nameKo: '처서', nameHanja: '處暑', nameEn: 'Chushu', angle: 150.0 },
  { key: 'bailu', nameKo: '백로', nameHanja: '白露', nameEn: 'Bailu', angle: 165.0 },
  { key: 'qiufen', nameKo: '추분', nameHanja: '秋分', nameEn: 'Qiufen', angle: 180.0 },
  { key: 'hanlu', nameKo: '한로', nameHanja: '寒露', nameEn: 'Hanlu', angle: 195.0 },
  { key: 'shuangjiang', nameKo: '상강', nameHanja: '霜降', nameEn: 'Shuangjiang', angle: 210.0 },
  { key: 'lidong', nameKo: '입동', nameHanja: '立冬', nameEn: 'Lidong', angle: 225.0 },
  { key: 'xiaoxue', nameKo: '소설', nameHanja: '小雪', nameEn: 'Xiaoxue', angle: 240.0 },
  { key: 'daxue', nameKo: '대설', nameHanja: '大雪', nameEn: 'Daxue', angle: 255.0 },
  { key: 'dongzhi', nameKo: '동지', nameHanja: '冬至', nameEn: 'Dongzhi', angle: 270.0 },
];

/**
 * JavaScript Date → 율리우스일(JD) 변환.
 * @param dt UTC 기준 Date 객체
 * @returns 율리우스일 (실수)
 */
export function jdFromDatetime(dt: Date): number {
  const seconds = dt.getTime() / 1000;
  return seconds / 86400.0 + 2440587.5;
}

/**
 * 율리우스일(JD) → JavaScript Date 변환.
 * @param jd 율리우스일 (실수)
 * @returns UTC 기준 Date 객체
 */
export function datetimeFromJd(jd: number): Date {
  const ms = (jd - 2440587.5) * 86400000;
  return new Date(ms);
}

/**
 * UTC 날짜·시간 요소로부터 율리우스일을 계산한다.
 * @param year 연도
 * @param month 월 (1~12)
 * @param day 일
 * @param hour 시
 * @param min 분
 * @param sec 초
 * @returns 율리우스일 (실수)
 */
export function jdFromUtcDate(
  year: number,
  month: number,
  day: number,
  hour: number,
  min: number,
  sec: number,
): number {
  const dt = new Date(Date.UTC(year, month - 1, day, hour, min, sec));
  if (year >= 0 && year < 100) dt.setUTCFullYear(year);
  return jdFromDatetime(dt);
}

/**
 * 특정 연도의 24절기 시점을 계산한다.
 *
 * 매일의 태양 황경을 추적하며, 목표 황경에 도달하는 시점을
 * 이분법(bisection)으로 정밀 산출한다.
 *
 * @param year 절기를 계산할 연도
 * @returns 24개 절기의 SolarTerm 배열
 */
export function computeSolarTerms(year: number): SolarTerm[] {
  const start = jdFromUtcDate(year, 1, 1, 0, 0, 0);
  const end = jdFromUtcDate(year + 1, 1, 1, 0, 0, 0);
  const days = Math.ceil(end - start);

  // 목표 황경을 단조증가로 정렬 (360° 경계 처리)
  const targets: number[] = [];
  let last = -1.0;
  for (const def of TERM_DEFS) {
    let angle = def.angle;
    while (angle <= last) {
      angle += 360.0;
    }
    targets.push(angle);
    last = angle;
  }

  const results: SolarTerm[] = [];
  let targetIdx = 0;
  let prevJd = start;
  let prevUnwrapped = sunApparentLongitude(prevJd);

  // 하루 단위로 태양 황경을 추적하며 절기 경계를 탐지
  for (let day = 1; day <= days; day++) {
    const jd = start + day;
    let lon = sunApparentLongitude(jd);
    if (lon < prevUnwrapped) {
      lon += 360.0;
    }
    while (targetIdx < targets.length && targets[targetIdx] <= lon) {
      const target = targets[targetIdx];
      if (target < prevUnwrapped) {
        targetIdx++;
        continue;
      }
      // 이분법으로 정밀 시점 산출
      const termJd = refineTerm(prevJd, jd, prevUnwrapped, target);
      results.push({ def: TERM_DEFS[targetIdx], jd: termJd });
      targetIdx++;
    }
    prevJd = jd;
    prevUnwrapped = lon;
  }

  return results;
}

/**
 * 태양 시황경(apparent longitude)을 계산한다.
 *
 * VSOP87 간략 공식 기반:
 * 1. 평균 황경(L0)과 평균 근점이각(M)을 계산
 * 2. 중심차(equation of center, C)를 적용하여 진황경 산출
 * 3. 장동(nutation) 보정을 적용하여 시황경 산출
 *
 * @param jd 율리우스일
 * @returns 태양 시황경 (0~360도)
 */
function sunApparentLongitude(jd: number): number {
  // 율리우스 세기(T) — J2000.0 에포크 기준
  const t = (jd - 2451545.0) / 36525.0;
  // 태양 평균 황경(L0)
  const l0 = 280.46646 + 36000.76983 * t + 0.0003032 * t * t;
  // 태양 평균 근점이각(M)
  const m = 357.52911 + 35999.05029 * t - 0.0001537 * t * t;
  const mRad = degToRad(m);
  // 중심차(C) — 케플러 방정식의 근사해
  const c =
    (1.914602 - 0.004817 * t - 0.000014 * t * t) * Math.sin(mRad) +
    (0.019993 - 0.000101 * t) * Math.sin(2.0 * mRad) +
    0.000289 * Math.sin(3.0 * mRad);
  const trueLong = l0 + c;
  // 장동(nutation) 보정
  const omega = 125.04 - 1934.136 * t;
  const lambda = trueLong - 0.00569 - 0.00478 * Math.sin(degToRad(omega));
  return normDeg(lambda);
}

/**
 * 이분법(bisection)으로 태양 황경이 목표값에 도달하는 JD를 정밀 산출한다.
 * 60회 반복으로 약 10⁻¹⁸일(≈ 0.1µs) 정밀도를 달성한다.
 *
 * @param jd0 탐색 시작 JD
 * @param jd1 탐색 종료 JD
 * @param lon0 jd0 시점의 태양 황경
 * @param target 목표 황경
 * @returns 목표 황경 도달 시점의 JD
 */
function refineTerm(jd0: number, jd1: number, lon0: number, target: number): number {
  let lo = jd0;
  let hi = jd1;
  let loLon = lon0;
  for (let i = 0; i < 60; i++) {
    const mid = (lo + hi) / 2.0;
    let midLon = sunApparentLongitude(mid);
    if (midLon < loLon) {
      midLon += 360.0;
    }
    if (midLon >= target) {
      hi = mid;
    } else {
      lo = mid;
      loLon = midLon;
    }
  }
  return (lo + hi) / 2.0;
}

/** 도(degrees) → 라디안(radians) 변환 */
function degToRad(deg: number): number {
  return (deg * Math.PI) / 180.0;
}

/** 각도를 0~360도 범위로 정규화 */
function normDeg(deg: number): number {
  deg = deg % 360.0;
  if (deg < 0.0) {
    deg += 360.0;
  }
  return deg;
}
